src/Components/Components/src/Routing/Router.cs

```diff
@@ -155,12 +181,55 @@ private void Refresh(bool isNavigationIntercepted)
             }
         }
 
+        private async Task RunOnNavigateAsync(string path, bool isNavigationIntercepted)
+        {
+            // If this router instance does not provide an OnNavigateAsync parameter
+            // then we render the component associated with the route as per usual.
+            if (OnNavigateAsync == null)
+            {
+                Refresh(isNavigationIntercepted);
+                return;
+            }
+
+            // Create a new CTS for this invocation of the task and
+            // extract the reference to the current one.
+            var pendingOnNavigateCts = _onNavigateCts;
+            var onNavigateCts = new CancellationTokenSource();
+            _onNavigateCts = onNavigateCts;
+
+            var navigateContext = new OnNavigateArgs(path, _onNavigateCts);
+            var task = OnNavigateAsync(navigateContext);
+
+            // If we've already invoked a task and stored its CTS, then
+            // cancel the existing task.
+            if (pendingOnNavigateCts != null)
+            {
+                pendingOnNavigateCts.Cancel();
+            }
+
+            // Create a cancellation task based on the cancellation token
+            // associated with the current running task.
+            var cancellationTaskSource = new TaskCompletionSource();
+            navigateContext.CancellationTokenSource.Token.Register(() =>
+                cancellationTaskSource.TrySetCanceled(navigateContext.CancellationTokenSource.Token));
+
+            // If the user provided a Loading render fragment, then show it.
+            if (Loading != null)
+            {
+                _renderHandle.Render(Loading);
+            }
+
+            await Task.WhenAny(task, cancellationTaskSource.Task);
+
+            Refresh(isNavigationIntercepted);
+        }
+
         private void OnLocationChanged(object sender, LocationChangedEventArgs args)
         {
             _locationAbsolute = args.Location;
             if (_renderHandle.IsInitialized && Routes != null)
             {
-                Refresh(args.IsNavigationIntercepted);
+                _ = RunOnNavigateAsync(NavigationManager.ToBaseRelativePath(_locationAbsolute), args.IsNavigationIntercepted);
```

Comment:
I think we need a way of flowing any errors from this task back into the renderer, otherwise any exceptions in `OnNavigateAsync` will be silently lost. We'd also want to be sure that if the `OnNavigateAsync`-returned `Task` gets cancelled, we don't treat that as an unhandled exception.

I fear this might involve creating some new APIs in `ComponentBase` and `Dispatcher` or something like that, since I don't immediately see any way to send an unhandled exception back into the renderer from outside.

If you want to ignore this issue for now and track it as a follow-up item (for either you or someone else on the team to handle before preview 8 ships), I think that would be totally fine, since it's conceptually a bit separate and I know this PR has already been expanding in scope quite a bit!

Reply:
@SteveSandersonMS This might be hacky, but we could probably flow the exception through InvokeAsync on the RenderHandle?

Reply:
I don't think that would be hacky at all - that would be totally fine... if it worked, which I don't think it will :)

The `InvokeAsync` on `Dispatcher` doesn't catch any exceptions from your task. It returns back a `Task` of its own that will throw if the underlying task throws.

So we might need some new API concept on the dispatcher like `InvokeAsync(task, throwOnFailure: true)`, and then the equivalent wrapper on `ComponentBase`. It's awkward which is why I'm suggesting handling it separately from this PR just to avoid expanding scope here too much.

Reply:
Fair enough, I don't know all the details so I would have to think about it. I'm fine if we put a pin on it so long as we file a tracking issue.

Reply:
Created a tracking issue: https://github.com/dotnet/aspnetcore/issues/23763